text/3729-sized-hierarchy.md

```diff
@@ -0,0 +1,1412 @@
+
+As `Sized` is to `const Sized`, `ValueSized` is to `const ValueSized`: `ValueSized`
+requires a value, knowledge of the type and target platform, and can only be
+computed at runtime. For example, `[svint8_t]` requires a value to know how
+many elements there are, and then information from the runtime environment
+to know the size of a `svint8_t`.
+
+`Pointee` is implemented by any type that can be used behind a pointer, which is
+to say, every type (put otherwise, these types may or may not be sized at all).
+For example, `Pointee` is therefore implemented on a `u32` which is trivially
+sized, a `[usize]` which is dynamically sized, a `svint8_t` which is
+runtime-sized and an `extern type` (from [rfcs#1861][rfc_extern_types]) which has
+no known size.
+
+All type parameters have an implicit bound of `const Sized` which will be
+automatically removed if a `Sized`, `const ValueSized`, `ValueSized` or
+`Pointee` bound is present instead.
+
+Prior to the introduction of `ValueSized` and `Pointee`, `Sized`'s implicit bound
+(now a `const Sized` implicit bound) could be removed using the `?Sized` syntax,
+which is now equivalent to a `ValueSized` bound in non-`const fn`s and
+`~const ValueSized` in `const fn`s and will be deprecated in the next edition.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Introduce a new marker trait, `ValueSized`, adding it to a trait hierarchy with the
+[`Sized`][api_sized] and [`Pointee`][api_pointee] traits, and make `Sized` and `ValueSized`
+traits `const`:
+
+```
+    ┌────────────────┐                  ┌─────────────────────────────┐
+    │ const Sized    │ ───────────────→ │ Sized                       │
+    │ {type, target} │     implies      │ {type, target, runtime env} │
+    └────────────────┘                  └─────────────────────────────┘
+            │                                          │
+         implies                                    implies
+            │                                          │
+            ↓                                          ↓
+┌───────────────────────┐             ┌────────────────────────────────────┐
+│ const ValueSized      │ ──────────→ │ ValueSized                         │
+│ {type, target, value} │   implies   │ {type, target, runtime env, value} │
+└───────────────────────┘             └────────────────────────────────────┘
+                                                       │
+                                                    implies
+                                                       │
+                               ┌───────────────────────┘
+                               ↓
+                     ┌──────────────────┐
+                     │ Pointee          │
+                     │ {runtime env, *} │
+                     └──────────────────┘
+```
+
+Or, in Rust syntax:
+
+```rust=
+const trait Sized: ~const ValueSized {}
```

Comment:
Side note: `const trait Trait {}` syntax is not a thing (yet), it's `#[const_trait] trait Trait {}` at the moment under `const_trait_impl`. Of course, this can stay as is if it's meant to be *conceptual*.

Reply:
Updated this to use the current syntax.